modules/light-clients/08-wasm/types/querier.go

```diff
@@ -0,0 +1,150 @@
+func NewQueryHandler(ctx sdk.Context, callerID string) *DefaultQuerier {
+	return &DefaultQuerier{
+		Ctx:      ctx,
+		CallerID: callerID,
+	}
+}
+
+// GasConsumed implements the wasmvmtypes.Querier interface.
+func (q *DefaultQuerier) GasConsumed() uint64 {
+	return VMGasRegister.ToWasmVMGas(q.Ctx.GasMeter().GasConsumed())
+}
+
+// Query implements the wasmvmtypes.Querier interface.
+func (q *DefaultQuerier) Query(request wasmvmtypes.QueryRequest, gasLimit uint64) ([]byte, error) {
+	sdkGas := VMGasRegister.FromWasmVMGas(gasLimit)
+
+	subCtx, _ := q.Ctx.WithGasMeter(storetypes.NewGasMeter(sdkGas)).CacheContext()
+
+	// make sure we charge the higher level context even on panic
+	defer func() {
+		q.Ctx.GasMeter().ConsumeGas(subCtx.GasMeter().GasConsumed(), "contract sub-query")
+	}()
+
+	if request.Stargate != nil {
+		return GetQueryPlugins().Stargate(subCtx, request.Stargate)
+	}
+
+	if request.Custom != nil {
+		return GetQueryPlugins().Custom(subCtx, request.Custom)
+	}
+
+	return nil, wasmvmtypes.UnsupportedRequest{Kind: "Unsupported query request"}
+}
+
+type (
+	CustomQuerier   func(ctx sdk.Context, request json.RawMessage) ([]byte, error)
+	StargateQuerier func(ctx sdk.Context, request *wasmvmtypes.StargateQuery) ([]byte, error)
+)
+
+// QueryPlugins is a list of query handlers that can be used to extend the default querier.
+type QueryPlugins struct {
+	Custom   CustomQuerier
+	Stargate StargateQuerier
+}
+
+// Merge merges the query plugin with a provided one.
+func (e QueryPlugins) Merge(x *QueryPlugins) QueryPlugins {
+	// only update if this is non-nil and then only set values
+	if x == nil {
+		return e
+	}
+
+	if x.Custom != nil {
+		e.Custom = x.Custom
+	}
+
+	if x.Stargate != nil {
+		e.Stargate = x.Stargate
+	}
+
+	return e
+}
+
+// SetQueryPlugins sets the current query plugins
+func SetQueryPlugins(plugins *QueryPlugins) {
+	queryPlugins = plugins
+}
+
+// GetQueryPlugins returns the current query plugins
+func GetQueryPlugins() *QueryPlugins {
+	return queryPlugins
+}
+
+// NewDefaultQueryPlugins returns the default set of query plugins
+func NewDefaultQueryPlugins() *QueryPlugins {
+	return &QueryPlugins{
+		Custom:   RejectCustomQuerier(),
+		Stargate: AcceptListStargateQuerier([]string{}),
+	}
+}
+
+// AcceptListStargateQuerier allows all queries that are in the accept list provided and in the default accept list.
+// This function returns protobuf encoded responses.
+func AcceptListStargateQuerier(accepted []string) func(sdk.Context, *wasmvmtypes.StargateQuery) ([]byte, error) {
+	return func(ctx sdk.Context, request *wasmvmtypes.StargateQuery) ([]byte, error) {
+		// A default list of accepted queries can be added here.
+		// accepted = append(defaultAcceptList, accepted...)
+
+		isAccepted := slices.Contains(accepted, request.Path)
+		if !isAccepted {
+			return nil, wasmvmtypes.UnsupportedRequest{Kind: fmt.Sprintf("'%s' path is not allowed from the contract", request.Path)}
```

Comment:
Should we use wasmvm errors or our own?

Reply:
wasmvm in my opinion. Similar to what wasmd does. This also helps with the `redactErr` function that I just added.